### ISSUE TITLE
[Tests] ShardSearchTransportRequestTests.testSerialization fails on 5.x

### ISSUE THREAD

**Issue by cbuescher** [elastic/elasticsearch#22808]
This fails reproducibly on 5.x (5d48757) as seen here: 

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.x+multijob-unix-compatibility/os=debian/474/consoleFull

Reproduce line:

```
gradle :core:test -Dtests.seed=CA62D2D3D29C115F -Dtests.class=org.elasticsearch.search.internal.ShardSearchTransportRequestTests -Dtests.method="testSerialization" -Dtests.security.manager=true -Dtests.locale=nl-BE -Dtests.timezone=Etc/GMT-14
```

**Comment by cbuescher** [comment]
Also failing on master (f128b7a7fe87486f633e235e125170c83dbd52b6):  

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=centos/557/consoleFull

```
gradle :core:test -Dtests.seed=A6318C18632167B8 -Dtests.class=org.elasticsearch.search.internal.ShardSearchTransportRequestTests -Dtests.method="testSerialization" -Dtests.security.manager=true -Dtests.locale=is -Dtests.timezone=Pacific/Yap
```

**Comment by cbuescher** [comment]
I traced this down to a different cacheKey in the deserialized ShardSearchTransportRequest. The reason seems to be that for the cacheKey computation we use ShardSearchLocalRequest#innerWriteTo which in turn deserialized the SearchSourceBuilder. While debugging I saw differences in the cacheKey bytes array where the source was rendered. Looking closer at the SearchSourceBuilder for this seed I found that the `source().collapse().getInnerHit()` renders to xContent in different ways in the order of the `scriptField` keys, because they are interenally stored as an unordered Set. Temporarily changing these to an ordered Set solved the problem for this particular seed.
